mne/preprocessing/xdawn.py

```diff
@@ -0,0 +1,565 @@
+    to any type of ERP.
+
+    Parameters
+    ----------
+    n_components : int, default 2
+        The number of components to decompose M/EEG signals.
+    signal_cov : None, Covariance instance or ndarray
+        The signal covariance used for whitening of the data. if None, the
+        covariance is estimated from the epochs signal.
+    correct_overlap : 'auto' or bool
+        Apply correction for overlaped ERP for the estimation of evokeds
+        responses. if 'auto', the overlapp correction is chosen in function
+        of the events in epochs.events.
+    reg : float, str, None
+        if not None, allow regularization for covariance estimation
+        if float, shrinkage covariance is used (0 <= shrinkage <= 1).
+        if str, optimal shrinkage using Ledoit-Wolf Shrinkage ('lws') or
+        Oracle Approximating Shrinkage ('oas').
+
+    Attributes
+    ----------
+    filters_ : dict of ndarray
+        If fit, the Xdawn components used to decompose the data for each event
+        type, else empty.
+    patterns_ : dict of ndarray
+        If fit, the Xdawn patterns used to restore M/EEG signals for each event
+        type, else empty.
+    evokeds_ : dict of evoked instance
+        If fit, the evoked response for each event type.
+
+    References
+    ----------
+    [1] Rivet, B., Souloumiac, A., Attina, V., & Gibert, G. (2009). xDAWN
+    algorithm to enhance evoked potentials: application to brain-computer
+    interface. Biomedical Engineering, IEEE Transactions on, 56(8), 2035-2043.
+
+    [2] Rivet, B., Cecotti, H., Souloumiac, A., Maby, E., & Mattout, J. (2011,
+    August). Theoretical analysis of xDAWN algorithm: application to an
+    efficient sensor selection in a P300 BCI. In Signal Processing Conference,
+    2011 19th European (pp. 1382-1386). IEEE.
+    """
+
+    def __init__(self, n_components=2, signal_cov=None, correct_overlap='auto',
+                 reg=None):
+        """init xdawn."""
+        self.n_components = n_components
+        self.signal_cov = signal_cov
+        self.reg = reg
+        self.filters_ = dict()
+        self.patterns_ = dict()
+        self.evokeds_ = dict()
+        self.projs_ = dict()
+
+        if correct_overlap not in ['auto', True, False]:
+            raise ValueError('correct_overlap must be a bool or "auto"')
+        self.correct_overlap = correct_overlap
+
+        # Extract signal covariance
+        if signal_cov is None:
+            self.signal_cov_ = None
+        elif isinstance(signal_cov, Covariance):
+            self.signal_cov_ = signal_cov.data
+        elif isinstance(signal_cov, np.ndarray):
+            self.signal_cov_ = signal_cov
+        else:
+            raise ValueError('signal_cov must be None, a covariance instance '
+                             'or a ndarray')
+
+    def fit(self, epochs, y=None):
+        """Fit Xdawn from epochs.
+
+        Parameters
+        ----------
+        epochs : Epoch object
```

Comment:
Impossible, xdawn need event position to work.
As discussed in this PR, and in issue #2245 , this code is more oriented denoising than decoding.
The XDAWN algorithm is not really compatible with the sklearn API where each sample are assumed to be unordered an independant.

You can still pipeline this in sklearn pipeline, but you have to do the CV by hand